backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/ReturnsInsertionLowering.kt

```diff
@@ -31,11 +32,17 @@ internal class ReturnsInsertionLowering(val context: Context) : FileLoweringPass
             override fun visitFunction(declaration: IrFunction) {
                 declaration.acceptChildrenVoid(this)
 
-                val body = declaration.body ?: return
-                if (declaration is IrConstructor || declaration.returnType == context.irBuiltIns.unitType) {
-                    val irBuilder = context.createIrBuilder(declaration.symbol, declaration.endOffset, declaration.endOffset)
-                    irBuilder.run {
-                         (body as IrBlockBody).statements += irReturn(irGetObject(symbols.unit))
+                context.createIrBuilder(declaration.symbol, declaration.endOffset, declaration.endOffset).run {
+                    when (val body = declaration.body) {
+                        is IrExpressionBody -> {
+                            declaration.body = IrBlockBodyImpl(body.startOffset, body.endOffset) {
+                                statements += irReturn(body.expression)
```

Comment:
What exactly is this required for?

Reply:
Some of the common lowerings produce `IrExpressionBody`, JVM/JS correctly handle it, so why wouldn't we do the same?

Reply:
I'm trying to ask: which lowerings? It is completely not clear for me how is this related to the main purpose of the PR. Does it enable these lowerings producing `IrExpressionBody`? Or does this patch fix unrelated bugs?

Reply:
SAM lowering, in particular (equals/hashCode from it, which weren't generated before this PR).